src/System.Private.CoreLib/shared/System/Text/AsciiUtility.cs

```diff
@@ -0,0 +1,1655 @@
+            {
+                Vector<ushort> maxAscii = new Vector<ushort>(0x007F);
+
+                if (Vector.LessThanOrEqualAll(Unsafe.ReadUnaligned<Vector<ushort>>(ref bufferInBytes), maxAscii))
+                {
+                    uint SizeOfVectorInBytes = (uint)Unsafe.SizeOf<Vector<ushort>>(); // JIT will make this a const
+
+                    // The first several elements of the input buffer were ASCII.
+                    // Pin and perform aligned reads until we find non-ASCII data or we hit the end of the buffer.
+                    // It's possible we'll end up reading the first part of the buffer twice; this is ok.
+
+                    bufferInBytes = ref Unsafe.AddByteOffset(ref bufferInBytes, SizeOfVectorInBytes);
+                    bufferLengthInBytes -= SizeOfVectorInBytes;
+
+                    fixed (byte* pBuffer = &bufferInBytes)
+                    {
+                        Debug.Assert(bufferLengthInBytes >= SizeOfVectorInBytes, "We should be able to perform at least one read.");
+
+                        byte* pFinalReadPos = pBuffer + bufferLengthInBytes - SizeOfVectorInBytes;
+                        byte* pAlignedBuffer = (byte*)((nuint)pBuffer & ~(nuint)(SizeOfVectorInBytes - 1));
+                        Debug.Assert(pFinalReadPos - pAlignedBuffer >= SizeOfVectorInBytes, "Bounds checking should've been performed earlier.");
+
+                        do
+                        {
+                            if (Vector.GreaterThanAny(Unsafe.Read<Vector<ushort>>(pAlignedBuffer), maxAscii))
+                            {
+                                break; // found non-ASCII data
+                            }
+                            pAlignedBuffer += SizeOfVectorInBytes;
+                        } while (pAlignedBuffer <= pFinalReadPos);
+
+                        nuint bytesConsumed = (nuint)pAlignedBuffer - (nuint)pBuffer;
+                        bufferLengthInBytes -= bytesConsumed;
+                        bufferInBytes = ref Unsafe.AsRef<byte>(pAlignedBuffer); // safe since the buffer is currently pinned
+                    }
+                }
+            }
+
+            // At this point, the buffer length wasn't enough to perform a vectorized search, or we did perform
+            // a vectorized search and encountered non-ASCII data. In either case go down a non-vectorized code
+            // path to drain any remaining ASCII chars.
+            //
+            // We're going to perform unaligned reads, so prefer 32-bit reads instead of 64-bit reads.
+            // This also allows us to perform more optimized bit twiddling tricks to count the number of ASCII chars.
+
+            uint currentUInt32;
+
+            // Try reading 64 bits at a time in a loop.
+            // From this point forward we don't need to keep track of the remaining buffer length.
+
+            {
```

Comment:
Is there a statement missing here or just extra curly braces?

Reply:
An earlier iteration of this code used a local whose name conflicted with a nested scope, but it has since been removed. I'll fix this up.